### ISSUE TITLE
5090 error

### ISSUE THREAD

**Issue by sunbin1993** [jax-ml/jax#30786]
### Description

code as below:

import jax
import jax.numpy as jnp
from jax import random, pmap, lax
from flax import linen as nn
import optax
from functools import partial

# 打印可用设备
print("可用设备:", jax.devices())
num_devices = len(jax.devices())

# 1. 定义模型
class MLP(nn.Module):
    @nn.compact
    def __call__(self, x):
        x = nn.Dense(512)(x)
        x = nn.relu(x)
        x = nn.Dense(256)(x)
        x = nn.relu(x)
        return nn.Dense(10)(x)

# 2. 初始化模型参数
def init_model(key):
    model = MLP()
    dummy_input = jnp.ones((1, 784))
    return model.init(key, dummy_input)

# 3. 定义损失函数
def loss_fn(params, batch):
    inputs, labels = batch
    preds = MLP().apply(params, inputs)
    return jnp.mean(optax.softmax_cross_entropy_with_integer_labels(preds, labels))

# 4. 定义优化器
optimizer = optax.adam(1e-3)

# 5. 数据并行训练步骤
@partial(pmap, axis_name='batch')
def train_step(params, opt_state, batch):
    inputs, labels = batch

    # 计算梯度和损失
    def step(params, inputs, labels):
        loss, grads = jax.value_and_grad(loss_fn)(params, (inputs, labels))
        return loss, grads

    loss, grads = step(params, inputs, labels)

    # 多卡梯度聚合
    grads = lax.pmean(grads, axis_name='batch')

    # 更新参数
    updates, opt_state = optimizer.update(grads, opt_state)
    params = optax.apply_updates(params, updates)
    return params, opt_state, loss

# 6. 主训练流程
def train():
    # 初始化
    key = random.PRNGKey(0)
    params = init_model(key)

    # 初始化优化器状态
    opt_state = optimizer.init(params)

    # JAX 0.7.0 的参数复制方式
    def replicate(tree):
        return jax.tree.map(lambda x: jnp.array([x] * num_devices), tree)

    replicated_params = replicate(params)
    replicated_opt_state = replicate(opt_state)

    # 模拟数据加载
    def get_batch():
        batch_size = 32
        key = random.PRNGKey(0)
        x = random.normal(key, (num_devices, batch_size, 784))
        y = random.randint(key, (num_devices, batch_size), 0, 10)
        return x, y

    # 训练循环
    for step in range(100):
        batch = get_batch()
        replicated_params, replicated_opt_state, loss = train_step(
            replicated_params, replicated_opt_state, batch
        )
        if step % 10 == 0:
            # JAX 0.7.0 的 device_get 方式
            print(f"Step {step}, Loss: {loss[0]:.4f}")

if __name__ == "__main__":
    train()

error as below:
E0806 07:42:41.619754    5613 pjrt_stream_executor_client.cc:2939] Execution of replica 0 failed: INTERNAL: NCCL operation ncclGroupEnd() failed: unhandled cuda error (run with NCCL_DEBUG=INFO for details). Last NCCL warning(error) log entry (may be unrelated) 'Cuda failure 'an illegal memory access was encountered''.
E0806 07:42:41.620149    5622 pjrt_stream_executor_client.cc:2939] Execution of replica 3 failed: INTERNAL: NCCL operation ncclGroupEnd() failed: unhandled cuda error (run with NCCL_DEBUG=INFO for details). Last NCCL warning(error) log entry (may be unrelated) 'Cuda failure 'an illegal memory access was encountered''.
F0806 07:42:51.619962    5148 pjrt_stream_executor_client.cc:3120] Replicated computation launch failed, but not all replicas terminated. Aborting process to work around deadlock. Failure message (there may have been multiple failures, see the error log for all failures): 

NCCL operation ncclGroupEnd() failed: unhandled cuda error (run with NCCL_DEBUG=INFO for details). Last NCCL warning(error) log entry (may be unrelated) 'Cuda failure 'an illegal memory access was encountered''.
*** Check failure stack trace: ***
    @     0x742e4c22e5c4  absl::lts_20250127::log_internal::LogMessage::SendToLog()
    @     0x742e4c22e396  absl::lts_20250127::log_internal::LogMessage::Flush()
    @     0x742e4c22ea39  absl::lts_20250127::log_internal::LogMessageFatal::~LogMessageFatal()
    @     0x742e43e1f4cd  xla::PjRtStreamExecutorLoadedExecutable::Execute()
    @     0x742e43d889e3  pjrt::PJRT_LoadedExecutable_Execute()
    @     0x74311060c146  xla::PjRtCApiLoadedExecutable::Execute()
    @     0x7431170b0efb  xla::ifrt::PjRtLoadedExecutable::Execute()
    @     0x743116b22fe1  xla::PyLoadedExecutable::ExecuteSharded()
    @     0x74310ff58a95  xla::ValueOrThrowWrapper<>::operator()()
    @     0x74310ff586ed  nanobind::detail::func_create<>()::{lambda()#1}::__invoke()
    @     0x74311706dd88  nanobind::detail::nb_func_vectorcall_complex()
    @           0x51e131  PyObject_Vectorcall
Aborted (core dumped)

### System info (python version, jaxlib version, accelerator, etc.)

| NVIDIA-SMI 570.153.02             Driver Version: 570.153.02     CUDA Version: 12.8     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 5090        Off |   00000000:16:00.0 Off |                  N/A |
| 35%   58C    P1            204W /  600W |    8225MiB /  32607MiB |     88%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA GeForce RTX 5090        Off |   00000000:27:00.0 Off |                  N/A |
|  0%   39C    P1             53W /  575W |     532MiB /  32607MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   2  NVIDIA GeForce RTX 5090        Off |   00000000:38:00.0 Off |                  N/A |
|  0%   35C    P1             43W /  575W |     534MiB /  32607MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   3  NVIDIA GeForce RTX 5090        Off |   00000000:98:00.0 Off |                  N/A |
|  0%   34C    P1             37W /  575W |     532MiB /  32607MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   4  NVIDIA GeForce RTX 5090        Off |   00000000:A8:00.0 Off |                  N/A |
|  0%   36C    P1             47W /  575W |     532MiB /  32607MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A            6246      C   python                                  528MiB |
|    1   N/A  N/A            6246      C   python                                  508MiB |
|    2   N/A  N/A            6246      C   python                                  508MiB |
|    3   N/A  N/A            6246      C   python                                  508MiB |
|    4   N/A  N/A            6246      C   python                                  508MiB |
+-----------------------------------------------------------------------------------------+

and jax version
jax                      0.7.0
jax-cuda12-pjrt          0.7.0
jax-cuda12-plugin        0.7.0
jaxlib                   0.7.0
jaxtyping                0.2.36

**Comment by Non1187** [comment]
Same error on NVIDIA GeForce RTX 2080 Ti. It could be a code error.

**Comment by sunbin1993** [comment]
> Same error on NVIDIA GeForce RTX 2080 Ti. It could be a code error.

fixed, i run  pip install torch==2.8.0

**Comment by Kun8018** [comment]
> > Same error on NVIDIA GeForce RTX 2080 Ti. It could be a code error.
> 
> fixed, i run pip install torch==2.8.0

Do you mean you install torch and run by torch? It means you run jax in 5090 still meet same error, right? And i wanna know your exactly jax version

**Comment by Kun8018** [comment]
@danielsuo We're hitting the same Illegal memory access error during multi-GPU training on a 5090 cluster using JAX 0.5.3. Could you confirm if there is a specific workaround or if we should migrate to the new PyTorch implementation?
